### PR TITLE
policy attachment: allow `targetRefs`

### PR DESCRIPTION
**Pull request by howardjohn**
Based on https://github.com/kubernetes-sigs/gateway-api/pull/2966. Note
that we do not HAVE to follow the GatewayAPI here; we can make our own
decision. There is, however, a general desire to allow multiple for
ergonomics.

In this proposal, I hide `targetRef`, but the API will remain + be
implemented forever. Implementation cost here is near zero, as we can
easily translate it to a single `targetRefs`; we just hide from docs to
push users toward the new ones.

### CODE REVIEW

**Comment by howardjohn** [comment]
cc @mikemorris @keithmattix

**Review by mikemorris — APPROVED** [review]
LGTM, is there any need to update the note below which is present on most of these fields to account for a singular `targetRet` in Istio 1.21?

```
  // NOTE: If you are using the `targetRefs` field in a multi-revision environment with Istio versions prior to 1.20,
  // it is highly recommended that you pin the authorization policy to a revision running 1.20+ via the istio.io/rev label.
  // This is to prevent proxies connected to older istiod control planes (that don't know about the targetRef field)
  // from misinterpreting the policy as namespace-wide during the upgrade process.
```

**Comment by howardjohn** [comment]
> LGTM, is there any need to update the note below which is present on most of these fields to account for a singular `targetRet` in Istio 1.21?
> 
> ```
>   // NOTE: If you are using the `targetRefs` field in a multi-revision environment with Istio versions prior to 1.20,
>   // it is highly recommended that you pin the authorization policy to a revision running 1.20+ via the istio.io/rev label.
>   // This is to prevent proxies connected to older istiod control planes (that don't know about the targetRef field)
>   // from misinterpreting the policy as namespace-wide during the upgrade process.
> ```

Yeah I think so (plus I think I botched some of the docs in the process of copy pasting so I'll clean up generally)

**Review by stevenctl — APPROVED** [review]

File: extensions/v1alpha1/wasm.proto [hunk 1]
```diff
@@ -246,17 +246,27 @@ message WasmPlugin {
...
 
+  // Optional. The targetRef specifies the gateway the policy should be
+  // applied to. The targeted resource specified will determine which
+  // workloads the policy applies to.
+  //
+  // Currently, the following resource attachment types are supported:
+  // * `kind: Gateway` with `group: gateway.networking.k8s.io` in the same namespace.
+  //
+  // If not set, the policy is applied as defined by the selector.
+  // At most one of the selector and targetRefs can be set.
```
Comment (ericvn) [inline extensions/v1alpha1/wasm.proto:260]: Nit: Highlighting is inconsistent, so my thought is to add it.
```suggestion
  // At most one of the `selector` and `targetRefs` can be set.
```

File: extensions/v1alpha1/wasm.proto [hunk 2]
```diff
@@ -246,17 +246,27 @@ message WasmPlugin {
   // At most, only one of the selector or `targetRef` can be set for a given policy.
```
Comment (ericvn) [inline extensions/v1alpha1/wasm.proto:246]: Should probably look at the other selector text as well to update to targetRefs.
```suggestion
  // At most one of the `selector` or `targetRefs` can be set for a given policy.
```

File: security/v1/authorization_policy.proto [hunk 3]
```diff
@@ -284,20 +284,26 @@ message AuthorizationPolicy {
   // and targetRef can be set.
```
Comment (ericvn) [inline security/v1/authorization_policy.proto:284]: Also prior line but I can't submit a change.
```suggestion
  // and `targetRefs` can be set.
```

File: security/v1/authorization_policy.proto [hunk 4]
```diff
@@ -284,20 +284,26 @@ message AuthorizationPolicy {
...
-  // must be a `Gateway` in the group `gateway.networking.k8s.io`. The
-  // gateway must be in the same namespace as the authorization policy.
+  // workloads the policy applies to.
+  //
+  // Currently, the following resource attachment types are supported:
+  // * `kind: Gateway` with `group: gateway.networking.k8s.io` in the same namespace.
   //
   // If not set, the policy is applied as defined by the selector.
-  // At most one of the selector and targetRef can be set.
+  // At most one of the selector and targetRefs can be set.
```
Comment (ericvn) [inline security/v1/authorization_policy.proto:298]:
```suggestion
  // At most one of the `selector` and `targetRefs` can be set.
```

File: security/v1beta1/authorization_policy.proto [hunk 5]
```diff
@@ -283,20 +283,26 @@ message AuthorizationPolicy {
...
-  // must be a `Gateway` in the group `gateway.networking.k8s.io`. The
-  // gateway must be in the same namespace as the authorization policy.
+  // workloads the policy applies to.
+  //
+  // Currently, the following resource attachment types are supported:
+  // * `kind: Gateway` with `group: gateway.networking.k8s.io` in the same namespace.
   //
   // If not set, the policy is applied as defined by the selector.
-  // At most one of the selector and targetRef can be set.
+  // At most one of the selector and targetRefs can be set.
```
Comment (ericvn) [inline security/v1beta1/authorization_policy.proto:297]:
```suggestion
  // At most one of the `selector` and `targetRefs` can be set.
```

File: security/v1beta1/request_authentication.proto [hunk 6]
```diff
@@ -254,17 +254,26 @@ message RequestAuthentication {
   // If not set, the selector will match all workloads. At most one of the selector and targetRef can be set.
```
Comment (ericvn) [inline security/v1beta1/request_authentication.proto:254]:
```suggestion
  // If not set, the `selector` will match all workloads. At most one of the `selector` and `targetRefs` can be set.
```

File: security/v1beta1/request_authentication.proto [hunk 7]
```diff
@@ -254,17 +254,26 @@ message RequestAuthentication {
...
+  // workloads the policy applies to.
+  //
+  // Currently, the following resource attachment types are supported:
+  // * `kind: Gateway` with `group: gateway.networking.k8s.io` in the same namespace.
   //
   // If not set, the policy is applied as defined by the selector.
-  // At most one of the selector and targetRef can be set.
-  // Waypoint proxies will not respect selectors even if they match.
-  istio.type.v1beta1.PolicyTargetReference targetRef = 3;
+  // At most one of the selector and targetRefs can be set.
```
Comment (ericvn) [inline security/v1beta1/request_authentication.proto:268]:
```suggestion
  // At most one of the `selector` and `targetRefs` can be set.
```

File: telemetry/v1/telemetry.proto [hunk 8]
```diff
@@ -264,23 +264,33 @@ option go_package = "istio.io/api/telemetry/v1";
...
+  // workloads the policy applies to.
+  //
+  // Currently, the following resource attachment types are supported:
+  // * `kind: Gateway` with `group: gateway.networking.k8s.io` in the same namespace.
   //
   // If not set, the policy is applied as defined by the selector.
-  // At most, only one of the selector or `targetRef` can be set for a given policy.
-  // Waypoint proxies will not respect selectors even if they match.
-  istio.type.v1beta1.PolicyTargetReference targetRef = 5;
+  // At most one of the selector and targetRefs can be set.
```
Comment (ericvn) [inline telemetry/v1/telemetry.proto:285]:
```suggestion
  // At most one of the `selector` and `targetRefs` can be set.
```

File: telemetry/v1alpha1/telemetry.proto [hunk 9]
```diff
@@ -262,23 +262,33 @@ option go_package = "istio.io/api/telemetry/v1alpha1";
...
+  // workloads the policy applies to.
+  //
+  // Currently, the following resource attachment types are supported:
+  // * `kind: Gateway` with `group: gateway.networking.k8s.io` in the same namespace.
   //
   // If not set, the policy is applied as defined by the selector.
-  // At most, only one of the selector or `targetRef` can be set for a given policy.
-  // Waypoint proxies will not respect selectors even if they match.
-  istio.type.v1beta1.PolicyTargetReference targetRef = 5;
+  // At most one of the selector and targetRefs can be set.
```
Comment (ericvn) [inline telemetry/v1alpha1/telemetry.proto:283]:
```suggestion
  // At most one of the `selector` and `targetRefs` can be set.
```

File: telemetry/v1alpha1/telemetry.proto [hunk 10]
```diff
@@ -262,23 +262,33 @@ option go_package = "istio.io/api/telemetry/v1alpha1";
...
 message Telemetry {
-  // Optional. The selector decides where to apply the Telemetry policy.
-  // If not set, the Telemetry policy will be applied to all workloads in the
-  // same namespace as the Telemetry policy.
+  // Optional. The selector decides where to apply the policy.
+  // If not set, the policy will be applied to all workloads in the
+  // same namespace as the policy.
   //
-  // At most, only one of the selector or targetRef can be set for a given policy.
+  // At most, only one of `selector` or `targetRefs` can be set for a given policy.
```
Comment (ericvn) [inline telemetry/v1alpha1/telemetry.proto:269]: Adding the for consistency. You could remove on the lines, but that would be more work.
```suggestion
  // At most, only one of the `selector` or `targetRefs` can be set for a given policy.
```

**Review by ericvn — COMMENTED** [review]
Nit on consistency with highlighting fields and including `the`.

**Review by ericvn — APPROVED** [review]

**Comment by stevenctl** [comment]
Did we miss WasmPlugin here?

**Comment by stevenctl** [comment]
and Telemetry

**Pull request merged**